rs/nns/integration_tests/src/why_list_neurons_expensive.rs

```diff
@@ -0,0 +1,417 @@
+use anyhow::bail;
+use candid::{Decode, Encode};
+use flate2::read::GzDecoder;
+use ic_base_types::PrincipalId;
+use ic_cdk::api::stable::WASM_PAGE_SIZE_IN_BYTES;
+use ic_nervous_system_string::clamp_debug_len;
+use ic_nns_constants::GOVERNANCE_CANISTER_ID;
+use ic_nns_governance_api::pb::v1::ListNeurons;
+use ic_nns_test_utils::state_test_helpers::{
+    list_neurons, primitive_get_profiling, unwrap_wasm_result, IcWasmTraceEvent,
+    IcWasmTraceEventType, PrimitiveIcWasmTraceEvent,
+};
+use ic_nns_test_utils_golden_nns_state::new_state_machine_with_golden_nns_state_or_panic;
+#[allow(unused)]
+use pretty_assertions::assert_eq;
+use std::{
+    collections::BTreeMap,
+    io::Read, // For flate2.
+    path::PathBuf,
+    str::FromStr,
+};
+// TODO: Figure out how to implement read_custom_sections using the walrus
+// library so that we are not using two WASM libraries.
+use wasmparser::{Parser, Payload};
+
+fn decompress_gz(buffer: &[u8]) -> Vec<u8> {
+    let mut result = vec![];
+    let mut decoder = GzDecoder::new(buffer);
+    decoder.read_to_end(&mut result).unwrap();
+    result
+}
+
+fn read_custom_section(wasm_bytes: &[u8], query_name: &str) -> Vec<u8> {
+    let parser = Parser::new(0);
+    let mut custom_section_data = None;
+
+    for payload in parser.parse_all(wasm_bytes) {
+        match payload.unwrap() {
+            Payload::CustomSection(custom_section) => {
+                if custom_section.name() == query_name {
+                    custom_section_data = Some(custom_section.data().to_vec());
+                    // This assumes that there are no more custom sections with
+                    // the same name.
+                    break;
+                }
+            }
+            _ => continue, // Ignore other payloads
+        }
+    }
+
+    custom_section_data.unwrap()
+}
+
+#[test]
+fn test_why_list_neurons_expensive() {
+    // Load golden nns state into a StateMachine.
+    let state_machine = new_state_machine_with_golden_nns_state_or_panic();
+
+    // Install governance WASM built from working copy. The main thing this does
+    // is allocate some stable memory that will be used to store profiling data
+    // that can be used to generate a flame graph. This space will be used by
+    // ic-wasm instrument.
+
+    println!("\nAllocating stable memory for profiling...\n");
```

Comment:
This is where we begin "instrumenting" the WASM, i.e. inject code that measures running cost. Measurements are saved during execution, and fetched afterwards. Then, they are fed into flamegraph generation.